### ISSUE TITLE
Simultaneous generation failed on SimulEval using MMA models

### ISSUE THREAD

**Issue by sarapapi** [pytorch/fairseq#3414]
Hi,
I've an issue regarding evaluation/generation using MMA trained model, in particular, I've updated my local branch today and when I try to use SimulEval I got this error:
```
/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torchaudio/backend/utils.py:53: UserWarning: "sox" backend is being deprecated. The default backend will be changed to "sox_io" backend in 0.8.0 and "sox" backend will be removed in 0.9.0. Please migrate to "sox_io" backend. Please refer to https://github.com/pytorch/audio/issues/903 for the detail.
  warnings.warn(
/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
2021-03-30 10:20:14 | INFO     | simuleval.scorer | Evaluating on speech
2021-03-30 10:20:14 | INFO     | simuleval.scorer | Source: /storage/MT/sara/simultaneous/dataset/MuST-Cinema/en-it/wav/amara_splitted/src_audiopath_list.txt
2021-03-30 10:20:14 | INFO     | simuleval.scorer | Target: /storage/MT/sara/simultaneous/dataset/MuST-Cinema/en-it/amara.it
2021-03-30 10:20:14 | INFO     | simuleval.scorer | Number of sentences: 545
2021-03-30 10:20:23 | INFO     | simuleval.server | Evaluation Server Started (process id 1773815). Listening to port 6667
2021-03-30 10:21:14 | WARNING  | simuleval.scorer | Resetting scorer
2021-03-30 10:21:14 | INFO     | simuleval.cli    | Output dir: /storage/MT/sara/simultaneous/training/ST/MMA_tag/
2021-03-30 10:21:14 | INFO     | simuleval.cli    | Evaluating FairseqSimulSTAgent (process id 1773808) on instances from 0 to 544
2021-03-30 10:21:14 | INFO     | simuleval.cli    | Start data writer (process id 1773855)
2021-03-30 10:23:27 | INFO     | fairseq.tasks.speech_to_text | dictionary size (spm_unigram10000_st.txt): 10,000
[W NNPACK.cpp:80] Could not initialize NNPACK! Reason: Unsupported hardware.
/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torchaudio/compliance/kaldi.py:574: UserWarning: The function torch.rfft is deprecated and will be removed in a future PyTorch release. Use the new torch.fft module functions, instead, by importing torch.fft and calling torch.fft.fft or torch.fft.rfft. (Triggered internally at  /pytorch/aten/src/ATen/native/SpectralOps.cpp:590.)
  fft = torch.rfft(strided_input, 1, normalized=False, onesided=True)
Traceback (most recent call last):
  File "/home/spapi/anaconda3/envs/fairseq/bin/simuleval", line 33, in <module>
    sys.exit(load_entry_point('simuleval', 'console_scripts', 'simuleval')())
  File "/home/spapi/fairseq_simul/SimulEval/simuleval/cli.py", line 165, in main
    _main(args.client_only)
  File "/home/spapi/fairseq_simul/SimulEval/simuleval/cli.py", line 192, in _main
    evaluate(args, client, server_process)
  File "/home/spapi/fairseq_simul/SimulEval/simuleval/cli.py", line 145, in evaluate
    decode(args, client, result_queue, indices)
  File "/home/spapi/fairseq_simul/SimulEval/simuleval/cli.py", line 108, in decode
    action = agent.policy(states)
  File "/home/spapi/fairseq_simul/fairseq/examples/speech_to_text/simultaneous_translation/agents/fairseq_simul_st_agent.py", line 325, in policy
    x, outputs = self.model.decoder.forward(
  File "/home/spapi/fairseq_simul/fairseq/fairseq/models/transformer.py", line 805, in forward
    x, extra = self.extract_features(
  File "/home/spapi/fairseq_simul/fairseq/examples/simultaneous_translation/models/transformer_monotonic_attention.py", line 219, in extract_features
    x, attn, _ = layer(
  File "/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/spapi/fairseq_simul/fairseq/examples/simultaneous_translation/modules/monotonic_transformer_layer.py", line 160, in forward
    x, attn = self.encoder_attn(
  File "/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/spapi/fairseq_simul/fairseq/examples/simultaneous_translation/modules/monotonic_multihead_attention.py", line 667, in forward
    p_choose = self.p_choose(
  File "/home/spapi/fairseq_simul/fairseq/examples/simultaneous_translation/modules/fixed_pre_decision.py", line 176, in p_choose
    p_choose_pooled = self.p_choose_waitk(
  File "/home/spapi/fairseq_simul/fairseq/examples/simultaneous_translation/modules/fixed_pre_decision.py", line 114, in p_choose_waitk
    p_choose = torch.tril(p_choose, diagonal=self.waitk_lagging - 1)
  File "/home/spapi/anaconda3/envs/fairseq/lib/python3.8/site-packages/torch/nn/modules/module.py", line 778, in __getattr__
    raise ModuleAttributeError("'{}' object has no attribute '{}'".format(
torch.nn.modules.module.ModuleAttributeError: 'MonotonicMultiheadAttentionInfiniteLookbackFixedStride' object has no attribute 'waitk_lagging'
```
@cndn @jmp84 @sravyapopuri388  I think that there's some mismatch between the previous pre-decision module and the actual one.
I add also that with the previous version of the code I scored a terrible BLEU score of 0.6 using MMA while a 14/15 using waitk.
Thanks

**Comment by Onoratofra** [comment]
Hi! I have the exact same issue.  The problem comes up in the training as well as the evaluation phase. 
Is there any chance that you will update the code so that we can perform comparative evaluation between waitk and MMA?

Thank you very much in advance!

`2021-04-21 09:31:38 | INFO | fairseq.trainer | begin training epoch 31
2021-04-21 09:31:38 | INFO | fairseq_cli.train | Start iterating over samples
Traceback (most recent call last):
  File "/home/fonorato/anaconda3/envs/myenv2/bin/fairseq-train", line 33, in <module>
    sys.exit(load_entry_point('fairseq', 'console_scripts', 'fairseq-train')())
  File "/home/fonorato/fairseqSLT/fairseq/fairseq_cli/train.py", line 489, in cli_main
    distributed_utils.call_main(cfg, main)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/distributed/utils.py", line 369, in call_main
    main(cfg, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq_cli/train.py", line 167, in main
    valid_losses, should_stop = train(cfg, trainer, task, epoch_itr)
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq_cli/train.py", line 277, in train
    log_output = trainer.train_step(samples)
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/trainer.py", line 662, in train_step
    loss, sample_size_i, logging_output = self.task.train_step(
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/tasks/fairseq_task.py", line 475, in train_step
    loss, sample_size, logging_output = criterion(model, sample)
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/criterions/label_smoothed_cross_entropy.py", line 79, in forward
    net_output = model(**sample["net_input"])
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/models/speech_to_text/convtransformer.py", line 221, in forward
    decoder_out = self.decoder(
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/fairseq/models/transformer.py", line 805, in forward
    x, extra = self.extract_features(
  File "/home/fonorato/fairseqSLT/fairseq/examples/simultaneous_translation/models/transformer_monotonic_attention.py", line 219, in extract_features
    x, attn, _ = layer(
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/examples/simultaneous_translation/modules/monotonic_transformer_layer.py", line 160, in forward
    x, attn = self.encoder_attn(
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/fonorato/fairseqSLT/fairseq/examples/simultaneous_translation/modules/monotonic_multihead_attention.py", line 667, in forward
    p_choose = self.p_choose(
  File "/home/fonorato/fairseqSLT/fairseq/examples/simultaneous_translation/modules/fixed_pre_decision.py", line 176, in p_choose
    p_choose_pooled = self.p_choose_waitk(
  File "/home/fonorato/fairseqSLT/fairseq/examples/simultaneous_translation/modules/fixed_pre_decision.py", line 114, in p_choose_waitk
    p_choose = torch.tril(p_choose, diagonal=self.waitk_lagging - 1)
  File "/home/fonorato/anaconda3/envs/myenv2/lib/python3.8/site-packages/torch/nn/modules/module.py", line 947, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'MonotonicMultiheadAttentionInfiniteLookbackFixedStride' object has no attribute 'waitk_lagging'`

**Comment by kurtisxx** [comment]
Hi @kahne @myleott , As you can see in this [issue](https://github.com/facebookresearch/SimulEval/issues/18), I have been trying to replicate results. I'd like to let you know that MMA code has not been working properly since forever. I am not even sure if it worked without any error even once.  I've been trying to fix this by the ppl you've assigned but they're not very responsive and helpful unfortunately.  As you can see in the issues, there are many others who have problems with replicating the results too.

**Comment by kurtisxx** [comment]
@MultiPath I am tagging you /reaching out to you regarding the "Monotonic Multihead Attention" work where you state that 

> "The code is available at https://github.com/pytorch/fairseq/tree/master/examples/simultaneous_translation."

 After spending weeks and big amount of money on this, I regret to inform you that I, like many others, couldn't replicate your results. I tried to fix this issue by getting in touch with the first author (@xutaima) of the paper but first author did nothing other than making me to spend more time as well as more money (told me that the command given in the README is not correct and I should train another model for much longer time) on this. It is not even possible to run MMA code currently due to the syntax errors in it (e.g. [PR](https://github.com/pytorch/fairseq/pull/3894)). Lately, @jmp84 involved in and stated that:

> "We're working towards a deadline so we don't have much bandwidth at the moment. In general, this kind of OSS project is provided as is with no guarantee of support. However, we will try to address issues on a best effort basis. Thank you for your understanding." ([source](https://github.com/facebookresearch/SimulEval/issues/18#issuecomment-928161193))

As I already said in my reply to this message, I am not looking for support from them to make MMA code compatible with current fairseq repository. I'd be more than happy if I am directed to any working combination of fairseq & MMA. One good thing about GitHub is that it allows you to reach out old versions of projects within seconds. So, it should not take more than 10 minutes to point out the versions of Fairseq and MMA which were working in the past. Don't you agree? Also, as you can see from the opening date of this issue, the problems in MMA are waiting to be resolved at least since April 21 which is the date @kahne assigned two person on this issue.

In summary, my question is very simple: Is there a code, as you stated in the paper, that can be used for replicating your bleu scores? If there is, could you please let me and others, who are interested in this, know where can we found it and how we can run it? If there is no code we can use to replicate, please also state it so that we will not spend more time on this. Thanks!

**Comment by xutaima** [comment]
Hi @kurtisxx, I really appreciate your interests in our work. However I really don't appreciate the pushing and uncomfortable way that you communicated, and some of your comments are pretty offensive. As the main maintainer of the MMA code, I am responsible for the codebase and I am regretful that the codebase brings inconvenience to you. But as @jmp84 has stated that we do have deadlines to work on at the moment. We do have plans to update code and documents in future, but I am afraid that would not fit your tight schedule. Thanks again for you interests and patience. I will follow up with you when there is update.

**Comment by FadedCosine** [comment]
@xutaima so have you updated the code till now?
